src/build_helper/src/git.rs

```diff
@@ -27,145 +28,196 @@ pub fn output_result(cmd: &mut Command) -> Result<String, String> {
     String::from_utf8(output.stdout).map_err(|err| format!("{err:?}"))
 }
 
-/// Finds the remote for rust-lang/rust.
-/// For example for these remotes it will return `upstream`.
-/// ```text
-/// origin  https://github.com/pietroalbani/rust.git (fetch)
-/// origin  https://github.com/pietroalbani/rust.git (push)
-/// upstream        https://github.com/rust-lang/rust (fetch)
-/// upstream        https://github.com/rust-lang/rust (push)
-/// ```
-pub fn get_rust_lang_rust_remote(
-    config: &GitConfig<'_>,
+/// Represents the result of checking whether a set of paths
+/// have been modified locally or not.
+#[derive(PartialEq, Debug, Clone)]
+pub enum PathFreshness {
+    /// Artifacts should be downloaded from this upstream commit,
+    /// there are no local modifications.
+    LastModifiedUpstream { upstream: String },
+    /// There are local modifications to a certain set of paths.
+    /// "Local" essentially means "not-upstream" here.
+    /// `upstream` is the latest upstream merge commit that made modifications to the
+    /// set of paths.
+    HasLocalModifications { upstream: String },
+    /// No upstream commit was found.
+    /// This should not happen in most reasonable circumstances, but one never knows.
+    MissingUpstream,
+}
+
+/// This function figures out if a set of paths was last modified upstream or
+/// if there are some local modifications made to them.
+/// It can be used to figure out if we should download artifacts from CI or rather
+/// build them locally.
+///
+/// The function assumes that at least a single upstream bors merge commit is in the
+/// local git history.
+///
+/// `target_paths` should be a non-empty slice of paths (relative to `git_dir` or the
+/// current working directory) whose modifications would invalidate the artifact.
+/// Each path can also be a negative match, i.e. `:!foo`. This matches changes outside
+/// the `foo` directory.
+///
+/// The function behaves differently in CI and outside CI.
+///
+/// - Outside CI, we want to find out if `target_paths` were modified in some local commit on
+/// top of the local master branch.
+/// If not, we try to find the most recent upstream commit (which we assume are commits
+/// made by bors) that modified `target_paths`.
+/// We don't want to simply take the latest master commit to avoid changing the output of
+/// this function frequently after rebasing on the latest master branch even if `target_paths`
+/// were not modified upstream in the meantime. In that case we would be redownloading CI
+/// artifacts unnecessarily.
+///
+/// - In CI, we always fetch only a single parent merge commit, so we do not have access
```

Comment:
```suggestion
/// - In CI, we use a sparse checkout. We fetch only a single parent merge commit, so we do not have access
```

Reply:
I'm not sure if we should call it a sparse checkout, tbh. Sparse checkout is a [separate git thing](https://git-scm.com/docs/git-sparse-checkout) that we are *not* using on CI. What you probably meant is a shallow clone, but that's also not accurate, because we actually checkout the last *two* commits. I'm not sure if there's a common term for that :)

Reply:
Ah, "shallow clone" is what I meant, yet. And I would say that is accurate, the clone is shallow. It has depth 1. That's not the same a depth 0, but it's still a shallow clone.
```suggestion
/// - In CI, we use a shallow clone of depth 1, i.e., we fetch only a single parent commit (which will be the most recent bors merge commit) and do not have access
```

Reply:
Ok, fair enough, "shallow" as a term is more general I suppose. This might be a bit confusing if someone takes a look at our [CI fetch code](https://github.com/rust-lang/rust/blob/aa4287cdb90d211c5eade007c1cd67ce4c268dee/.github/workflows/ci.yml#L91), which confusingly uses depth 2 (because depth 0 here means "fetch everything"). But that's a small thing. Used your text to hopefully clarify the comment.

Reply:
If the clone has depth 2 we should say that, "1" was just a guess on my side. :)

Reply:
Actually, yes, the clone depth is indeed 2. I thought that shallow clone is `--depth=0`, but I remembered that wrong.

Reply:
"2-depth checkout"